File: app/src/main/java/danb/speedrunbrowser/utils/AppDatabase.java
package danb.speedrunbrowser.utils;


import android.content.Context;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.Dao;
import androidx.room.Database;
import androidx.room.Delete;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.PrimaryKey;
import androidx.room.Query;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import androidx.room.migration.Migration;
import androidx.sqlite.db.SupportSQLiteDatabase;
import danb.speedrunbrowser.BuildConfig;
import io.reactivex.Completable;
import io.reactivex.Maybe;
import io.reactivex.Single;

@Database(entities = {AppDatabase.WatchHistoryEntry.class, AppDatabase.Subscription.class}, version = 3)
public abstract class AppDatabase extends RoomDatabase {

    public static final Migration MIGRATION_1_2 = new Migration(1, 2) {
        @Override
        public void migrate(@NonNull SupportSQLiteDatabase database) {
            database.execSQL("CREATE TABLE IF NOT EXISTS `Subscription` (`resourceId` TEXT NOT NULL, `type` TEXT NOT NULL, PRIMARY KEY(`resourceId`))");
        }
    };

    public static final Migration MIGRATION_2_3 = new Migration(2, 3) {
        @Override
        public void migrate(@NonNull SupportSQLiteDatabase database) {
            database.execSQL("ALTER TABLE `Subscription` ADD COLUMN `name` TEXT NOT NULL DEFAULT '';");
        }
    };

    public static AppDatabase make(Context ctx) {
        return Room.databaseBuilder(ctx, AppDatabase.class, "appdb")
                .addMigrations(
                        MIGRATION_1_2,
                        MIGRATION_2_3
                )
                .build();
    }

    public abstract WatchHistoryDao watchHistoryDao();
    public abstract SubscriptionDao subscriptionDao();

    @Entity
    public static class Subscription {
        Subscription() {}

        @Ignore
        public Subscription(@NonNull String type, @NonNull String resourceId, @NonNull String name) {
            this.resourceId = resourceId;
            this.type = type;
            this.name = name;
        }

        @PrimaryKey
        @NonNull
        public String resourceId;

        @NonNull
        public String type;

        @NonNull
        public String name;

        public String getFCMTopic() {
            if(BuildConfig.DEBUG) {
                return "debug_" + type + "_" + resourceId;
            }
            else
                return "release_" + type + "_" + resourceId;
        }

        @Override
        public boolean equals(@Nullable Object obj) {
            return obj instanceof Subscription && ((Subscription)obj).resourceId.equals(resourceId);
        }

        @Override
        public int hashCode() {
            return resourceId.hashCode();
        }
    }

    @Entity
    public static class WatchHistoryEntry {

        WatchHistoryEntry(@NonNull String runId) {
            this.runId = runId;
        }

        @Ignore
        public WatchHistoryEntry(@NonNull String runId, long seekPos) {
            this.runId = runId;

            this.watchDate = System.currentTimeMillis();

            this.seekPos = seekPos;
        }

        @PrimaryKey
        @NonNull
        public String runId;

        public long watchDate;
        public long seekPos;
    }

    @Dao
    public interface SubscriptionDao {
        @Query("SELECT * FROM Subscription WHERE resourceId = :resourceId")
        Maybe<Subscription> get(String resourceId);

        @Query("SELECT * FROM Subscription WHERE type = :type ORDER BY name LIMIT 40 OFFSET :offset")
        Single<List<Subscription>> listOfType(String type, int offset);

        @Query("SELECT * FROM Subscription WHERE type = :type AND resourceId LIKE :idPrefix || '%'")
        Single<List<Subscription>> listOfTypeWithIDPrefix(String type, String idPrefix);

        @Query("SELECT * FROM Subscription WHERE type = :type AND name LIKE :filter ORDER BY name LIMIT 40 OFFSET :offset")
        Single<List<Subscription>> listOfTypeWithFilter(String type, String filter, int offset);

        @Insert(onConflict = OnConflictStrategy.REPLACE)
        Completable subscribe(Subscription... subscriptions);

        @Delete
        Completable unsubscribe(Subscription... subscriptions);
    }

    @Dao
    public interface WatchHistoryDao {
        @Query("SELECT * FROM WatchHistoryEntry ORDER BY watchDate DESC LIMIT 40 OFFSET :offset")
        Single<List<WatchHistoryEntry>> getMany(int offset);

        @Query("SELECT * FROM WatchHistoryEntry WHERE runId = :runId")
        Maybe<WatchHistoryEntry> get(String runId);

        @Insert(onConflict = OnConflictStrategy.REPLACE)
        Completable record(WatchHistoryEntry... historyEntries);

        @Delete
        Completable delete(WatchHistoryEntry historyEntry);
    }
}

File: app/src/main/java/danb/speedrunbrowser/utils/AppDatabase.kt
package danb.speedrunbrowser.utils


import android.content.Context
import androidx.room.Dao
import androidx.room.Database
import androidx.room.Delete
import androidx.room.Entity
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import danb.speedrunbrowser.BuildConfig
import io.reactivex.Completable
import io.reactivex.Maybe
import io.reactivex.Single

@Database(entities = [AppDatabase.WatchHistoryEntry::class, AppDatabase.Subscription::class], version = 3)
abstract class AppDatabase : RoomDatabase() {

    abstract fun watchHistoryDao(): WatchHistoryDao
    abstract fun subscriptionDao(): SubscriptionDao

    @Entity
    data class Subscription(
            var type: String,
            @PrimaryKey var resourceId: String,
            var name: String
    ) {

        val fcmTopic: String
            get() = if (BuildConfig.DEBUG) {
                "debug_" + type + "_" + resourceId
            } else
                "release_" + type + "_" + resourceId

        override fun equals(other: Any?): Boolean {
            return other is Subscription && other.resourceId == resourceId
        }

        override fun hashCode(): Int {
            return resourceId.hashCode()
        }
    }

    @Entity
    data class WatchHistoryEntry(
        @PrimaryKey
        var runId: String,
        var seekPos: Long = 0
    ) {
        var watchDate: Long = System.currentTimeMillis()
    }

    @Dao
    interface SubscriptionDao {
        @Query("SELECT * FROM Subscription WHERE resourceId = :resourceId")
        operator fun get(resourceId: String): Maybe<Subscription>

        @Query("SELECT * FROM Subscription WHERE type = :type ORDER BY name LIMIT 40 OFFSET :offset")
        fun listOfType(type: String, offset: Int): Single<List<Subscription>>

        @Query("SELECT * FROM Subscription WHERE type = :type AND resourceId LIKE :idPrefix || '%'")
        fun listOfTypeWithIDPrefix(type: String, idPrefix: String): Single<List<Subscription>>

        @Query("SELECT *, substr(origResourceId, 1, pos - 1) as resourceId FROM (SELECT *, instr(resourceId, '_') as pos, resourceId as origResourceId FROM Subscription WHERE type = :type) GROUP BY substr(origResourceId, 1, pos - 1) ORDER BY name LIMIT 40 OFFSET :offset")
        fun listOfTypeGrouped(type: String, offset: Int): Single<List<Subscription>>

        @Query("SELECT * FROM Subscription WHERE type = :type AND name LIKE :filter ORDER BY name LIMIT 40 OFFSET :offset")
        fun listOfTypeWithFilter(type: String, filter: String, offset: Int): Single<List<Subscription>>

        @Insert(onConflict = OnConflictStrategy.REPLACE)
        fun subscribe(vararg subscriptions: Subscription): Completable

        @Delete
        fun unsubscribe(vararg subscriptions: Subscription): Completable
    }

    @Dao
    interface WatchHistoryDao {
        @Query("SELECT * FROM WatchHistoryEntry ORDER BY watchDate DESC LIMIT 40 OFFSET :offset")
        fun getMany(offset: Int): Single<List<WatchHistoryEntry>>

        @Query("SELECT * FROM WatchHistoryEntry WHERE runId = :runId")
        operator fun get(runId: String): Maybe<WatchHistoryEntry>

        @Insert(onConflict = OnConflictStrategy.REPLACE)
        fun record(vararg historyEntries: WatchHistoryEntry): Completable

        @Delete
        fun delete(historyEntry: WatchHistoryEntry): Completable
    }

    companion object {

        private val MIGRATION_1_2: Migration = object : Migration(1, 2) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("CREATE TABLE IF NOT EXISTS `Subscription` (`resourceId` TEXT NOT NULL, `type` TEXT NOT NULL, PRIMARY KEY(`resourceId`))")
            }
        }

        private val MIGRATION_2_3: Migration = object : Migration(2, 3) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("ALTER TABLE `Subscription` ADD COLUMN `name` TEXT NOT NULL DEFAULT '';")
            }
        }

        fun make(ctx: Context): AppDatabase {
            return Room.databaseBuilder(ctx, AppDatabase::class.java, "appdb")
                    .addMigrations(
                            MIGRATION_1_2,
                            MIGRATION_2_3
                    )
                    .build()
        }
    }
}
